### PR TITLE
Chunk explorer

### PR DESCRIPTION
**Pull request by holisticode**
This PR aims at introducing the possibility to explore chunk references.

It adds a `explore` command for the `main` binary. This would print all hashes (alphabetically sorted) for a given file to STDOUT

The idea is that the command be used to generate a list of hashes.

This is all up for discussion and not meant to be opened upstream yet.

### CODE REVIEW

**Comment by nonsense** [comment]
@holisticode I think we should split this PR in two:
1. the `explore` part seems fine to me and we can probably merge soon. Isn't `swarm references` or `swarm hashes` a better name, rather than `explore`?
2. the `chunk-explore` does not belong in `swarm-smoke` IMO and also it is not efficient to start a new process for every single chunk. The same way `explore` returns a list of many references, we should have a command that checks many references, not just one, and returns structured output.

File: cmd/swarm/explore.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		utils.Fatalf("Error opening file " + args[1])
+	}
+	defer f.Close()
+
+	fileStore := storage.NewFileStore(&storage.FakeChunkStore{}, storage.NewFileStoreParams())
+	refs, err := fileStore.GetAllReferences(context.TODO(), f, false)
+	if err != nil {
+		utils.Fatalf("%v\n", err)
+	} else {
+		for _, r := range refs {
```
Comment (nonsense) [inline cmd/swarm/explore.go:55]: `refs` should be sorted - either here or in `GetAllReferences`.

**Comment by holisticode** [comment]
> @holisticode I think we should split this PR in two:

> the explore part seems fine to me and we can probably merge soon. Isn't swarm references or swarm hashes a better name, rather than explore?

Agreed, so I understand you are suggesting to use `references` or `hashes` as command extension to the `swarm` binary right?

> the chunk-explore does not belong in swarm-smoke IMO and also it is not efficient to start a new process for every single chunk. 

Agree here too. I used the `swarm-smoke` binary as a short-cut to reuse the functionality of setting up endpoints, cluster and nodes to the query. But as we can't use `ws` for this currently anyway, it has no reason to stay there at all.

> The same way explore returns a list of many references, we should have a command that checks many references, not just one, and returns structured output.

Also agree. Are you saying as a command extension to the `swarm` binary? Separate binary? What is your suggestion to query multiple nodes (all nodes in a cluster namespace for example)?

**Comment by nonsense** [comment]
@holisticode 
> Agreed, so I understand you are suggesting to use references or hashes as command extension to the swarm binary right?

Yes, one of these two names is better than `explore` IMO :)


> Agree here too. I used the swarm-smoke binary as a short-cut to reuse the functionality of setting up endpoints, cluster and nodes to the query. But as we can't use ws for this currently anyway, it has no reason to stay there at all.

We should modify the `swarm-smoke` to take a list of `hosts` instead of generating them based on `cluster, from, to, scheme, etc.` and then we can target the `ws` endpoint from within the k8s cluster.

> Also agree. Are you saying as a command extension to the swarm binary? Separate binary? What is your suggestion to query multiple nodes (all nodes in a cluster namespace for example)?

RPC call, like `hasChunks` that you can query over `ws` would do the trick. Then you can iterate over all `ws` endpoints and check all nodes.

File: cmd/swarm/swarm-smoke/main.go [hunk 2]
```diff
@@ -88,6 +88,12 @@ func main() {
 			Usage:       "http or https",
 			Destination: &scheme,
 		},
+		cli.StringFlag{
```
Comment (nonsense) [inline cmd/swarm/swarm-smoke/main.go]: Not needed.

File: cmd/swarm/swarm-smoke/main.go [hunk 3]
```diff
@@ -160,6 +166,12 @@ func main() {
 			Usage:   "measure network aggregate capacity",
 			Action:  wrapCliCommand("sliding-window", false, slidingWindow),
 		},
+		{
+			Name:    "chunk_explorer",
```
[the remark this answers is not in the document]
Comment (janos) [reply cmd/swarm/swarm-smoke/main.go]: Ok, that is fine, too.

File: cmd/swarm/explore.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/ethereum/go-ethereum/cmd/utils"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+	"gopkg.in/urfave/cli.v1"
+)
+
+var chunkExploreCommand = cli.Command{
```
Comment (nonsense) [inline cmd/swarm/explore.go]: `hashesCommand`

File: cmd/swarm/explore.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	Name:               "hashes",
+	Usage:              "print all hashes of a file to STDOUT",
+	ArgsUsage:          "<file>",
+	Description:        "Prints all hashes of a file to STDOUT",
+}
+
+func hashes(ctx *cli.Context) {
+	args := ctx.Args()
+	if len(args) < 1 {
+		utils.Fatalf("Usage: swarm explore <file name>")
```
Comment (nonsense) [inline cmd/swarm/explore.go]: `swarm hashes`

File: swarm/storage/filestore.go [hunk 6]
```diff
@@ -96,3 +97,42 @@ func (f *FileStore) Store(ctx context.Context, data io.Reader, size int64, toEnc
 func (f *FileStore) HashSize() int {
 	return f.hashFunc().Size()
 }
+
+// Public API. This endpoint returns all chunk hashes (only) for a given file
```
Comment (nonsense) [inline swarm/storage/filestore.go]: Godoc begins with the name of the function, so:

```
GetAllReferences is a public API endpoint, which returns all chunk hashes for a given file.
```

File: swarm/storage/filestore_test.go [hunk 7]
```diff
@@ -173,3 +173,39 @@ func testFileStoreCapacity(toEncrypt bool, t *testing.T) {
...
+	}
+	db := tdb.LDBStore
+	memStore := NewMemStore(NewDefaultStoreParams(), db)
+	localStore := &LocalStore{
+		memStore: memStore,
+		DbStore:  db,
+	}
+	fileStore := NewFileStore(localStore, NewFileStoreParams())
+
+	checkRefs := func(dataSize int, expectedLen int) {
```
Comment (nonsense) [inline swarm/storage/filestore_test.go]: No need for this function, just unnecessary indirection. Just inline it on line 209. https://github.com/golang/go/wiki/TableDrivenTests

**Comment by nonsense** [comment]
@holisticode the description says:
```
a `chunk-explore` command option for the `swarm-smoke` binary., which queries the `has` RPC endpoint of nodes if the locally store a certain chunk. This second command has to be refined yet and further discussed, as it is not possible currently to query RPC endpoints via `ws` at the moment for cluster deployments. Maybe a different way to query nodes .
```

It is possible to query cluster deployments via `ws` so maybe update this part when you submit upstream.

**Review by nonsense — APPROVED** [review]

**Review by janos — APPROVED** [review]

**Comment by holisticode** [comment]
Reopened upstream as https://github.com/ethereum/go-ethereum/pull/19008

**Pull request closed without merging**